recipes/TicTacToeReactHooks/src/Main.purs

```diff
@@ -0,0 +1,277 @@
+module TicTacToeReactHooks.Main where
+
+import Prelude
+
+import Control.Alt ((<|>))
+import Data.Array (concat, (:))
+import Data.Array as Array
+import Data.Array.NonEmpty (NonEmptyArray)
+import Data.Array.NonEmpty as NonEmpty
+import Data.Foldable (or)
+import Data.Generic.Rep (class Generic)
+import Data.Generic.Rep.Eq (genericEq)
+import Data.Generic.Rep.Show (genericShow)
+import Data.Int (even)
+import Data.Maybe (Maybe(..), fromMaybe, isJust)
+import Effect (Effect)
+import Effect.Exception (throw)
+import React.Basic.DOM (CSS, css, render)
+import React.Basic.DOM as R
+import React.Basic.Events (EventHandler, handler_)
+import React.Basic.Hooks (Component, JSX, Reducer, component, keyed, mkReducer, useReducer, (/\))
+import React.Basic.Hooks as React
+import Web.HTML (window)
+import Web.HTML.HTMLDocument (body)
+import Web.HTML.HTMLElement (toElement)
+import Web.HTML.Window (document)
+
+main :: Effect Unit
+main = do
+  body <- body =<< document =<< window
+  case body of
+    Nothing -> throw "Could not find body."
+    Just b -> do
+      game <- mkGame
+      render (game unit) (toElement b)
+      mempty
+
+mkSquare :: Component { onClick :: EventHandler, value :: Square }
+mkSquare =
+  component "Square" \props -> React.do
+    let
+      text = case props.value of
+        Just player -> show player
+        Nothing -> ""
+    pure
+      $ R.button
+          { style: styles.square
+          , onClick: props.onClick
+          , children: [ R.text text ]
+          }
+
+type Square
+  = Maybe Player
+
+data Player
+  = X
+  | O
+
+derive instance genericPlayer :: Generic Player _
+
+instance showPlayer :: Show Player where
+  show = genericShow
+
+instance eqPlayer :: Eq Player where
+  eq = genericEq
+
+mkBoard ::
+  Component
+    { onClick :: Int -> Int -> EventHandler
+    , squares :: BoardState
+    }
+mkBoard = do
+  square <- mkSquare
+  component "Board" \props -> React.do
+    let
+      renderSquare i j = square { value: props.squares i j, onClick: props.onClick i j }
+    pure
+      $ R.div_
+          [ R.div
+              { style: styles.boardRow
+              , children: [ 0, 1, 2 ] <#> renderSquare 0
+              }
+          , R.div
+              { style: styles.boardRow
+              , children: [ 0, 1, 2 ] <#> renderSquare 1
+              }
+          , R.div
+              { style: styles.boardRow
+              , children: [ 0, 1, 2 ] <#> renderSquare 2
+              }
+          ]
+
+data Action
+  = JumpToStep Int
+  | FillSquare Int Int
+
+type BoardState
+  = Int -> Int -> Square
```

Comment:
I'm not sure how much of this is worth commenting on, but this is a notable departure from the original example in the React docs, where board state is an array of `'X' | 'O' | null`. Instead of initializing an `Array (Maybe Player)` full of `Nothing` and then having to handle the nested `Maybe (Maybe Player)` when indexing into that array, I thought this representation was a bit easier to work with in PureScript.